kong/dao/migrations/helpers.lua

```diff
@@ -95,5 +99,213 @@ function _M.plugin_config_iterator(dao, plugin_name)
+
+        if source_value ~= nil then
+          len = len + 1
+          values[len] = source_value
+          column_names[len] = dest_column_name
+        end
+      end
+
+      local question_marks = string.sub(string.rep("?, ", len), 1, -3)
+
+      local insert_cql = fmt("INSERT INTO %s (%s) VALUES (%s)",
+                             destination_table_def.name,
+                             table_concat(column_names, ", "),
+                             question_marks)
+      local _, err = coordinator:execute(insert_cql, values, CASSANDRA_EXECUTE_OPTS)
+      if err then
+        return nil, err
+      end
+    end
+  end
+end
+
+do
+
+  local function create_table(coordinator, table_def)
+
+    local partition_keys = table_def.partition_keys
+    local primary_key_cql = ""
+    if #partition_keys > 0 then
+      primary_key_cql = fmt(", PRIMARY KEY (%s)", table_concat(partition_keys, ", "))
+    end
+
+    local column_declarations = {}
+    local len = 0
+    for name, typ in pairs(table_def.columns) do
+      len = len + 1
+      column_declarations[len] = fmt("%s %s", name, typ)
+    end
+    local column_declarations_cql = table_concat(column_declarations, ", ")
+
+    local cql = fmt("CREATE TABLE %s(%s%s);",
+                    table_def.name,
+                    column_declarations_cql,
+                    primary_key_cql)
+    return coordinator:execute(cql, {}, CASSANDRA_EXECUTE_OPTS)
+  end
+
+
+  local function drop_table_if_exists(coordinator, table_name)
+    local cql = fmt("DROP TABLE IF EXISTS %s;", table_name)
+    return coordinator:execute(cql, {}, CASSANDRA_EXECUTE_OPTS)
+  end
+
+
+  local function get_columns_to_copy(table_structure)
+    local res = {}
+    for k, _ in pairs(table_structure.columns) do
+      res[k] = k
+    end
+    return res
+  end
+
+
+  local function create_aux_table_def(table_def)
+    local aux_table_def = utils.deep_copy(table_def)
+    aux_table_def.name = "copy_of_" .. table_def.name
+    aux_table_def.columns.partition = "text"
+    table.insert(aux_table_def.partition_keys, 1, "partition")
+    return aux_table_def
+  end
+
+  --[[
+    Add a new partition key called "partition" to the table specified by table_def.
+
+    table_def has the following structure:
+    { name    = "ssl_certificates",
+      columns = {
+        id         = "uuid",
+        cert       = "text",
+        key        = "text",
+        created_at = "timestamp",
+      },
+      partition_keys = { "id" },
+    }
+  --]]
+
+  function _M.cassandra.add_partition(dao, table_def)
+
+    local coordinator, err = dao.db:get_coordinator()
+    if not coordinator then
+      return nil, err
+    end
+
+    table_def = utils.deep_copy(table_def)
+
+    local aux_table_def = create_aux_table_def(table_def)
+    local columns_to_copy = get_columns_to_copy(table_def)
+    columns_to_copy.partition = function() return cassandra.text(table_def.name) end
+
+    local _, err = create_table(coordinator, aux_table_def)
```

Comment:
This helper needs to be able to re-run if it failed halfway through in a previous attempt, just in case. I think this operation should be `create_table_if_exists`, otherwise it would never succeed a re-run, right?

We should fix all other similar places which prevent this migration from being re-run imho.

Reply:
Thanks! I've done the change (`create_table_if_not_exists` instead of `create_table_if_exists`).

> This helper needs to be able to re-run

This made me think. If the helper collapses in the middle of the record copying, it will be difficult to recuperate it (the records would have to be removed from the destination database, otherwise they would be duplicated). I see two ways of approaching this:

* Deleting all records from the destination table before copying the records.
* Not do the initial delete, and instead each time a record is copied from source to destination, the record is then removed from source.

The first option seems more conservative because it delays erasing stuff. The second option seems faster.

But there is a catch: if the first option fails precisely because there is "too much data" (i.e. the memory on the coordinator is spent, or too many operations provoke a timeout) then running the migration again won't help, because every time it has to copy all the records. If the second option fails for the same reason, when it is run again the records will already be partially copied, so it is more likely to succeed in case of "too much data" errors.

For this reason I am partial to the second option. But maybe I am overthinking this. Please let me know your thoughts.